### ISSUE TITLE
Report diagnostic for unnecessary null check

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#900]
When a readonly/init-only field or property is initialized with a non-null value, a null check is not necessary.
Would be great if you could report a diagnostic for this, but I'm not sure it's possible.

Example:
```c#
using System;

internal sealed class Program
{
    public static string Name { get; } = "Meziantou";
    private readonly string _address = "ABC";

    public void Run(string[] args)
    {
        if (Name is null)
        {
            throw new InvalidOperationException($"{nameof(Name)} must not be null");
        }

        if (_address is null)
        {
            throw new InvalidOperationException($"{nameof(_address)} must not be null");
        }
    }
}
```

I stumbled upon this [here](https://github.com/meziantou/Meziantou.Analyzer/blob/c5ff3361ddf02ef10381770a646c8e4f6e4eea30/tests/Meziantou.Analyzer.Test/Helpers/ProjectBuilder.Validation.cs#L19-L22) by accident.

**Comment by meziantou** [comment]
This one is more complex. Because of reflection and all the scenarios that use it (e.g. deserialization), you can find many cases where the check may be useful. Only `static readonly` would be _safe_ to handle.

Also, note that not all projects use nullable reference types and the compiler doesn't report case where you use a null check on a non-null value. So, you cannot always rely on these annotations.

In terms of performance, `static readonly` are handled correctly by the JIT and the condition should be elided (at least, in tier 1).

For completeness, the rule should handle
- `value == null`, `null == value` when the `==` operator is not defined on the type (otherwise the behavior may not be a null check)
- pattern matching `is null`
- `ReferenceEquals(value, null)`, `ReferenceEquals(null, value)`
In each case, you have to check if value is 
- a field or property reference
    - is static readonly 
    - Has an initializer
    - The value must be non-null (may not be trivial to validate) or the member must be annotated as not nullable.

So, it doesn't seem to be a trivial rule and may affect very few scenarios.

**Comment by drieseng** [comment]
I agree. Feel free to close this one.

Sorry for the noise :-)
